Framework/NexusGeometry/src/NexusGeometrySave.cpp

```diff
@@ -0,0 +1,1038 @@
+    writeOrientation(transformations, compInfo, index, locationIsOrigin);
+  }
+
+  H5::StrType dependencyStrType = strTypeOfSize(dependency);
+  writeXYZPixeloffset(childGroup, compInfo, index);
+  writeNXDetectorNumber(childGroup, compInfo, detIds, index);
+
+  writeStrDataset(childGroup, LOCAL_NAME, name);
+  writeStrDataset(childGroup, DEPENDS_ON, dependency);
+}
+
+/*
+ * Function: Saveinstrument
+ * writes the instrument from memory to disk in Nexus format.
+ *
+ * @param compInfo : componentInfo object.
+ * @param fullPath : save destination as full path.
+ * @param reporter : report to progressBase.
+ */
+void saveInstrument(
+    const std::pair<std::unique_ptr<Geometry::ComponentInfo>,
+                    std::unique_ptr<Geometry::DetectorInfo>> &instrPair,
+    const std::string &fullPath, const std::string &rootPath,
+    Kernel::ProgressBase *reporter) {
+
+  const Geometry::ComponentInfo &compInfo = (*instrPair.first);
+  const Geometry::DetectorInfo &detInfo = (*instrPair.second);
+
+  // Exception handling.
+  boost::filesystem::path tmp(fullPath);
+  if (!boost::filesystem::is_directory(tmp.root_directory())) {
+    throw std::invalid_argument(
+        "The path provided for saving the file is invalid: " + fullPath + "\n");
+  }
+  const auto ext = boost::filesystem::path(tmp).extension();
+  if ((ext != ".nxs") && (ext != ".hdf5")) {
+    throw std::invalid_argument("invalid extension for file: " +
+                                ext.generic_string());
+  }
+  if (reporter != nullptr) {
+    reporter->report();
+  }
+  if (!compInfo.hasDetectorInfo()) {
+    throw std::invalid_argument("The component has no detector info.\n");
+  }
+  if (!compInfo.hasSample()) {
+
+    throw std::invalid_argument("The component has no sample.\n");
+  }
+
+  if (Mantid::Kernel::V3D{0, 0, 0} != compInfo.samplePosition()) {
+    throw std::invalid_argument("The sample is not at the origin.\n");
+  }
+
+  if (!compInfo.hasSource()) {
+    throw std::invalid_argument("The component has no source.");
+  }
+
+  // open file
+  H5::H5File file(fullPath, H5F_ACC_TRUNC); // open file
+
+  std::vector<size_t> nxDetectors = nxDetectorIndices(compInfo);
+  std::vector<size_t> nxMonitors = nxMonitorIndices(detInfo);
+  H5::Group rootGroup, instrument;
+
+  // create NXentry (file root)
+  rootGroup = file.createGroup(rootPath);
+  NexusGeometrySave::writeStrAttribute(rootGroup, NX_CLASS, NX_ENTRY);
+
+  // save and capture NXinstrument (component root)
+  instrument = NexusGeometrySave::NXInstrument(rootGroup, compInfo);
+
+  // save NXdetectors
+  for (const size_t &index : nxDetectors) {
+    NexusGeometrySave::saveNXDetector(instrument, compInfo, detInfo, index);
+  }
+
+  // save NXmonitors
+  for (const size_t &index : nxMonitors) {
+    NexusGeometrySave::saveNXMonitor(instrument, compInfo, detInfo, index);
+  }
+
+  // save NXsource
+  NexusGeometrySave::saveNXSource(instrument, compInfo);
```

Comment:
I'm think you should write out sample and source before any monitors or detectors. See comment above about master loop.

Reply:
done